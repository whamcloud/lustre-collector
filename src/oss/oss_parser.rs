// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{
    oss::{ldlm_parser, obdfilter_parser},
    types::Record,
};
use combine::{choice, error::ParseError, Parser, Stream};

pub(crate) fn params() -> Vec<String> {
    let mut a = obdfilter_parser::obd_params();
    a.extend(ldlm_parser::ldlm_params());

    a
}

pub(crate) fn parse<I>() -> impl Parser<I, Output = Record>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    choice((obdfilter_parser::parse(), ldlm_parser::parse()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use combine::many;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_params() {
        let x = r#"obdfilter.fs-OST0000.stats=
snapshot_time             1535148988.363769785 secs.nsecs
write_bytes               9 samples [bytes] 98303 4194304 33554431
create                    4 samples [reqs]
statfs                    42297 samples [reqs]
get_info                  2 samples [reqs]
connect                   6 samples [reqs]
reconnect                 1 samples [reqs]
disconnect                4 samples [reqs]
statfs                    46806 samples [reqs]
preprw                    9 samples [reqs]
commitrw                  9 samples [reqs]
ping                      8229 samples [reqs]
obdfilter.fs-OST0000.num_exports=2
obdfilter.fs-OST0000.tot_dirty=0
obdfilter.fs-OST0000.tot_granted=8666816
obdfilter.fs-OST0000.tot_pending=0
"#;

        let result: (Vec<_>, _) = many(parse()).parse(x).unwrap();

        assert_debug_snapshot!(result)
    }
}
