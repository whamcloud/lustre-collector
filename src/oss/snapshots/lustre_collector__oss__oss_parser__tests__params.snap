---
source: src/oss/oss_parser.rs
expression: result
---
(
    [
        Target(
            Stats(
                TargetStat {
                    kind: Ost,
                    param: Param(
                        "stats",
                    ),
                    target: Target(
                        "fs-OST0000",
                    ),
                    value: [
                        Stat {
                            name: "write_bytes",
                            units: "bytes",
                            samples: 9,
                            min: Some(
                                98303,
                            ),
                            max: Some(
                                4194304,
                            ),
                            sum: Some(
                                33554431,
                            ),
                            sumsquare: None,
                        },
                        Stat {
                            name: "create",
                            units: "reqs",
                            samples: 4,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "statfs",
                            units: "reqs",
                            samples: 42297,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "get_info",
                            units: "reqs",
                            samples: 2,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "connect",
                            units: "reqs",
                            samples: 6,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "reconnect",
                            units: "reqs",
                            samples: 1,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "disconnect",
                            units: "reqs",
                            samples: 4,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "statfs",
                            units: "reqs",
                            samples: 46806,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "preprw",
                            units: "reqs",
                            samples: 9,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "commitrw",
                            units: "reqs",
                            samples: 9,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "ping",
                            units: "reqs",
                            samples: 8229,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                    ],
                },
            ),
        ),
        Target(
            NumExports(
                TargetStat {
                    kind: Ost,
                    param: Param(
                        "num_exports",
                    ),
                    target: Target(
                        "fs-OST0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            TotDirty(
                TargetStat {
                    kind: Ost,
                    param: Param(
                        "tot_dirty",
                    ),
                    target: Target(
                        "fs-OST0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            TotGranted(
                TargetStat {
                    kind: Ost,
                    param: Param(
                        "tot_granted",
                    ),
                    target: Target(
                        "fs-OST0000",
                    ),
                    value: 8666816,
                },
            ),
        ),
        Target(
            TotPending(
                TargetStat {
                    kind: Ost,
                    param: Param(
                        "tot_pending",
                    ),
                    target: Target(
                        "fs-OST0000",
                    ),
                    value: 0,
                },
            ),
        ),
    ],
    "",
)
